onnxruntime/core/providers/cuda/object_detection/non_max_suppression.cc

```diff
@@ -0,0 +1,134 @@
+    for (int64_t class_index = 0; class_index < pc.num_classes_; ++class_index) {
+      IAllocatorUniquePtr<void> d_selected_indices{};
+      IAllocatorUniquePtr<void> h_number_selected_ptr{AllocateBufferOnCPUPinned<void>(sizeof(int))};
+      auto* h_number_selected = static_cast<int*>(h_number_selected_ptr.get());
+
+      ORT_RETURN_IF_ERROR(NonMaxSuppressionImpl(
+          [this](size_t bytes) { return GetScratchBuffer<void>(bytes); },
+          pc,
+          GetCenterPointBox(),
+          batch_index,
+          class_index,
+          max_output_boxes_per_class,
+          iou_threshold,
+          score_threshold,
+          d_selected_indices,
+          h_number_selected));
+
+      int num_saved_outputs = *h_number_selected;
+      if (num_saved_outputs > 0) {
+        all_selected_indices.emplace_back(std::move(d_selected_indices), num_saved_outputs);
+        total_num_saved_outputs += num_saved_outputs;
+      }
+    }
+  }
+
+  if (total_num_saved_outputs == 0) {
+    ctx->Output(0, {0, 3});
+  } else {
+    // concatenate outputs
+    const auto last_dim = 3;
+    const auto num_elements = last_dim * total_num_saved_outputs;
+    Tensor* output = ctx->Output(0, {static_cast<int64_t>(total_num_saved_outputs), last_dim});
+    ORT_ENFORCE(output != nullptr);
+    int64_t* dst = output->MutableData<int64_t>();
+    size_t count = all_selected_indices.size();
+
+    CudaAsyncBuffer<const void*> input_ptr(this, count);
+    CudaAsyncBuffer<int64_t> concat_sizes_gpu(this, count);
+    CudaAsyncBuffer<int64_t> concat_sizes_range_gpu(this, count);
+    CudaAsyncBuffer<int64_t> axis_dimension_input_output_mapping_gpu(this, total_num_saved_outputs);
+
+    int index = 0;
+    for (size_t i = 0; i < count; i++) {
+      auto& it = all_selected_indices[i];
+      auto src = std::get<0>(it).get();
+      auto size = std::get<1>(it);
+
+      input_ptr.CpuPtr()[i] = src;
+      concat_sizes_gpu.CpuPtr()[i] = size;
+      concat_sizes_range_gpu.CpuPtr()[i] = (i == 0) ? size : size + concat_sizes_range_gpu.CpuPtr()[i - 1];
+      for (int j = 0; j < size; j++) {
+        axis_dimension_input_output_mapping_gpu.CpuPtr()[index++] = i;
+      }
+    }
+
+    concat_sizes_gpu.CopyToGpu();
+    axis_dimension_input_output_mapping_gpu.CopyToGpu();
+    concat_sizes_range_gpu.CopyToGpu();
+    input_ptr.CopyToGpu();
+
+    ORT_RETURN_IF_ERROR(ConcatImpl(sizeof(int64_t),
+                                   num_elements,
+                                   last_dim,
+                                   concat_sizes_gpu.GpuPtr(),
+                                   concat_sizes_range_gpu.GpuPtr(),
+                                   axis_dimension_input_output_mapping_gpu.GpuPtr(),
+                                   count,
+                                   dst,
+                                   input_ptr.GpuPtr(),
+                                   num_elements));
+  }
```

Comment:
I think this part mostly similar with the code in Concat. Is it possible to has a shared function.